constraints_lang.md

```diff
@@ -0,0 +1,311 @@
+
+As an example, we'll try to construct a sketch of a polygon where dependencies flow in both directions.
+
+Note that the code is generated by the user clicking in the UI. The resulting code is in the order that it was created, meaning it's a history of actions that the user performed in the UI.
+
+```
+p1 = screen_point(x: 10, y: 12)
+p2 = screen_point(x: 20, y: 13)
+l1 = line(p1, p2)
+p3 = screen_point(x: 21, y: 23)
+l2 = line(p2, p3)
+p4 = screen_point(x: 21, y: 23)
+l3 = line(p3, p4)
+l4 = line(p4, p1)
+||l3|| = 0.5 * ||l1||^2
+||l2|| = 1.2 * ||l4||
+```
+
+The above code can be read as follows:
+
+1. User enters sketch mode, which defaults to line drawing.
+2. User clicks a point. A variable `p1` is defined with the screen coordinates.
+3. User clicks another point. A variable `p2` is defined at that point. Since they're in line-sketch mode, `l1 = line(p1, p2)` is also generated. The app knows to name it with the prefix `l` because it is generating the line, so it knows it's a line. Similarly with points.
+4. Repeat the above for points `p3` and `p4`.
+5. When they close the sketch by clicking `p1`, the app recognizes this, does not define a fifth point, and creates `l4 = line(p4, p1)`, closing the sketch.
+6. The user selects one of the lines, clicks "Constrain length with function". This opens an input box in the command palette where the user can type an arbitrary expression.
+7. The user types `0.5 * ||`, and the UI knows that the user wants to select a piece of geometry.
+8. The user selects a line with their pointing device, and the app fills in `l1||`.
+9. The user types `^2 Enter`, which finishes the expression.
+10. At this point, the app appends at the end of the code the new line `||l3|| = 0.5 * ||l1||^2`.
+	- Design note: We could add this relationship next to the definition of `l3`. This is a reasonable choice. But I think we should prefer to append to the end by default because it retains the history of user operations in the order that they did them, which can be useful when reading code. Not only do UI operations correspond one-to-one with lines of code (a spatial relationship), but they correspond in order to how the author created them (a temporal relationship). This strengthens the tie between the UI and code in the user's mind. If they find that they'd like a different order, they can reorder lines without affecting the meaning of the code in many cases. (We'll discuss exactly which cases later.)
```

Comment:
The idea that anyone but power users would know to type  `||` into the KCL expression box to initiate a selection seems a bit far-fetched to me. I would rather a UX where we add `len(l1)` (see my note above) on selection automatically, and if the user wants something other than magnitude they'd have to specify. But I think I understand this whole example flow as a deeply power user one anyway, and most would use simpler constraint applications like `parellel` anyway.